Wikipedia/Code/MWKDataStore.m

```diff
@@ -988,7 +987,59 @@ - (NSArray *)imageInfoForArticleWithURL:(NSURL *)url {
 
 #pragma mark - helper methods
 
-- (void)iterateOverArticles:(void (^)(MWKArticle *))block {
+- (void)removeUnreferencedArticlesFromDiskCacheWithFailure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
+    NSFetchRequest *allValidArticleKeysFetchRequest = [WMFArticle fetchRequest];
+    allValidArticleKeysFetchRequest.predicate = [NSPredicate predicateWithFormat:@"viewedDate != NULL || savedDate != NULL"];
+    allValidArticleKeysFetchRequest.propertiesToFetch = @[@"key"];
+
+    NSError *fetchError = nil;
+    NSArray *arrayOfAllValidArticles = [self.viewContext executeFetchRequest:allValidArticleKeysFetchRequest error:&fetchError];
+
+    if (fetchError) {
+        failure(fetchError);
+        return;
+    }
+
+    if (arrayOfAllValidArticles.count == 0) {
+        [[NSFileManager defaultManager] removeItemAtPath:[self pathForSites] error:nil];
```

Comment:
Might be good to move this onto the background queue though in case it hangs long enough to cause an `8badf00d`